docs/book/v3/features/container/config.md

```diff
@@ -0,0 +1,213 @@
+{
+    return new SomePageHandler(
+        $container->get(TemplateRendererInterface::class)
+    );
+}
+```
+
+You can, however, re-use a factory for multiple services by accepting the second
+argument and varying creation based on it:
+
+```php
+use Psr\Container\ContainerInterface;
+use Zend\Expressive\Template\TemplateRendererInterface;
+
+class PageFactory
+{
+    public function __invoke(ContainerInterface $container, string $serviceName)
+    {
+        $name = strtolower($serviceName);
+        return new PageHandler(
+            $container->get(TemplateRendererInterface::class),
+            $name
+        );
+    };
+}
+```
+
+The above could be mapped for several services:
+
+```php
+return [
+    'dependencies' => [
+        'factories' => [
+            'hello-world' => PageFactory::class,
+            'about'       => PageFactory::class,
+        ],
+    ],
+];
+```
+
+In general, services should be cached by the container after initial creation;
+factories should only be called once for any given service name.
+
+## Invokables
+
+_Invokables_ refer to any class that may be instantiated without any constructor
+arguments. In other words, one should be able to create an instance solely be
+calling `new $className()`.
+
+Configuration for invokables looks verbose; it's a map of the service name to
+the class name to instantiate, and, generally, these are the same values.
+
+However, you can _also_ provide a different service name. In those situations,
+containers MUST treat the service name as an alias to the final class name, and
+allow retrieving the service by EITHER the alias OR the class name.
+
+As an example, given the following configuration:
+
+```php
+'dependencies' => [
+    'invokables' => [
+        'HelloWorld' => PageAction::class,
+    ],
+],
+```
+
+the container should allow retrieval of both the services "HelloWorld" as well
+as the "PageAction" class.
+
+## Delegator Factories
+
+Delegator factories are factories that may be used to _decorate_ or _manipulate_
+a service before returning it from the container. They are covered in detail [in
+another chapter](delegator-factories.md), and delegator factories have the
+following signature:
+
+```php
+use Psr\Container\ContainerInterface;
+
+function (
+    ContainerInterface $container,
+    string $serviceName,
+    callable $callback
+)
+```
+
+Configuration for delegator factories is using the "delegators" sub-key of the
+"dependencies" configuration. Each entry is a service name pointing to an
+_array_ of delegator factories. Delegator factories are called in the order they
+appear in configuration; when you call the `$callback` argument, it effectively
+calls the previous delegator to get the instance to work on; the first
+delegator's `$callback` argument invokes whatever functionality is required to
+create the initial instance (be it as a [service](#services),
+[invokable](#invokables), or [factory](#factories), after [alias](#aliases)
+resolution).
```

Comment:
Not sure if I understand it correctly, please see tests:
https://github.com/webimpress/zend-container-test/blob/master/src/DelegatorTestTrait.php

Reply:
Your tests are (mostly) correct. (One thing to note: you have a number of tests supposedly testing that delegators are running when retrieving aliased services, but, in fact, you're calling `get()` on the non-aliased versions.)

When you call `get()` for a service that has delegator factories defined, you will receive the value returned by the last delegator in the list. The first delegator's `$callback()` is essentially the result of calling `get()` as if there were no delegators in place.

I'll see if I can reword this paragraph to make it more clear.